src/applications/train/electra.py

```diff
@@ -0,0 +1,284 @@
+from collections.abc import Mapping
+from typing import Any, Dict, List, Union
+
+import torch
+from datasets import load_dataset
+from torch import nn
+from transformers import (
+    DataCollatorForLanguageModeling,
+    ElectraConfig,
+    ElectraForMaskedLM,
+    ElectraForPreTraining,
+    ElectraTokenizerFast,
+    PretrainedConfig,
+    PreTrainedModel,
+    Trainer,
+)
+from transformers.data.data_collator import _torch_collate_batch
+
+# # 3. ELECTRA (replaced token detection objective)
+# see details in paper [ELECTRA: Pre-training Text Encoders as Discriminators Rather Than Generators](https://arxiv.org/abs/2003.10555)
+
+
+class ELECTRAModel(PreTrainedModel):
+    def __init__(self, generator, discriminator, config):
+        super().__init__(config)
+        self.generator, self.discriminator = generator, discriminator
+        self.gumbel_dist = torch.distributions.gumbel.Gumbel(0.0, 1.0)
+        self.pad_token_id = config.pad_token_id
+
+    def to(self, *args, **kwargs):
+        "Also set dtype and device of contained gumbel distribution if needed"
+        super().to(*args, **kwargs)
+        a_tensor = next(self.parameters())
+        device, dtype = a_tensor.device, a_tensor.dtype
+        dtype = torch.float32
+        self.gumbel_dist = torch.distributions.gumbel.Gumbel(
+            torch.tensor(0.0, device=device, dtype=dtype),
+            torch.tensor(1.0, device=device, dtype=dtype),
+        )
+
+    # N.B.: removed "sentA_lenths" -- didn't figure out what it was for
```

Comment:
Can you pass a link to the code implementation and I can take a look to see if I can figure it out